### PR TITLE
Add RAFT model for optical flow

### PR DESCRIPTION
**Pull request by NicolasHug**
Towards https://github.com/pytorch/vision/issues/4644

This PR adds the RAFT model, along with its basic building blocks (feature encoder, correlation block, update block, etc.) and two model builder function: `raft_large()` and `raft_small()`.

The architecture (not the code!) is exactly the same as the original implementation from https://github.com/princeton-vl/RAFT, which will allow us support the original paper's weights if we want to. This architecture differs slightly from what is described in the paper. I have annotated the paper with these differences, hoping this can help the review:
[RAFT- Recurrent All-Pairs Field Transforms forOptical Flow (1).pdf](https://github.com/pytorch/vision/files/7641423/RAFT-.Recurrent.All-Pairs.Field.Transforms.forOptical.Flow.1.pdf)

A summary is this:

image

----

### API

The `RAFT` class accepts **torch.nn.Module instances** as input and offers a low-level API.

```py
RAFT(
        feature_encoder=feature_encoder,
        context_encoder=context_encoder,
        corr_block=corr_block,
        update_block=update_block,
        mask_predictor=mask_predictor,
    )
```
The model builder functions `raft_large()` and `raft_small()` are higher-level and do not require any parameter. They can however take as input the same parameters as the `RAFT` class, so as to override their defaults. E.g.:

```py
model = raft_large(mask_predictor=MyOwnMaskPredictor())  # all the other building blocks will use the default architecture
```

The building blocks like `FeatureEncoder`, `ResidualBlock`,  `UpdateBlock`, etc. are (sort of publicly) available in `torchvision.models.optical_flow.raft`, but are not exposed in `__all__`.

----

Still left TODO, here or in follow-up PRs:

- [x] Add an expect test for both `raft()` and `raft_small()`
- [x] Add better docs
- [ ] Add type annotations (🥲)

In follow up PRs I will also submit the training reference and associated transforms.

I will write review comments below to hightlight important bits, or things where I'm not to sure what the best way is.

cc @datumbox

### CODE REVIEW

File: torchvision/models/optical_flow/__init__.py
Comment (NicolasHug) [inline torchvision/models/optical_flow/__init__.py:1]: We should expose the building blocks as well, stuff like `ResidualBlock`, `FeatureEncoder`, etc.

Should we expose these in a `models.optical_flow.raft` namespace?

I remember @datumbox mentioning a few issues when we have both a module and a function with the same name. Ideally, I'd like to keep the `raft()` name for the function builder, but I'm happy to get your thoughts on this

File: torchvision/models/optical_flow/_raft/raft.py
Comment (NicolasHug) [inline torchvision/models/optical_flow/_raft/raft.py:12]: We can probably merge this with the resnet implementation (same for BottleneckBlock below)... but this might make the API awkward and bloated. For now I'd say it's simpler and safer to keep these implementations separate

File: torchvision/models/optical_flow/_raft/raft.py
Comment (NicolasHug) [inline torchvision/models/optical_flow/_raft/raft.py:151]: Does it make sense to use a `ConvNormActivation` while passing `norm_layer=None`? No strong opinion from me

File: torchvision/models/optical_flow/_raft/raft.py
Comment (NicolasHug) [inline torchvision/models/optical_flow/_raft/raft.py:424]: kwargs are here to override the `RAFT` class parameters, e.g. to override the entire `feature_encoder`, or the whole `update_block`.

File: torchvision/models/optical_flow/_raft/raft.py
Comment (NicolasHug) [inline torchvision/models/optical_flow/_raft/raft.py:97]: I named this `layers` to be sort of consistent with the `ResNet` class. I'm happy to consider other names though

File: torchvision/models/optical_flow/_raft/raft.py
Comment (NicolasHug) [inline torchvision/models/optical_flow/_raft/raft.py:191]: Where possible I tried to write defaults that will correspond to the normal RAFT model (kernel_size, padding). I refrained from doing that for the input and output shapes though. Thoughts?

File: torchvision/models/optical_flow/_raft/raft.py
Comment (NicolasHug) [inline torchvision/models/optical_flow/_raft/raft.py:246]: I'm keeping the 0.25 default for consistency with the original code, but I'm tempted to set it to 1 to encourage users not to use it

File: torchvision/models/optical_flow/_raft/raft.py
Comment (NicolasHug) [inline torchvision/models/optical_flow/_raft/raft.py:367]: This is a bit that I did not see mentioned in the paper. I wish we could separate the initialization of the hidden state from the context encoder, but this would likely prevent us from exactly reproduce the original implementation.

File: torchvision/models/optical_flow/_raft/raft.py
Comment (NicolasHug) [inline torchvision/models/optical_flow/_raft/raft.py:354]: This "8" downscaling factor is hard-coded in different places. We should be able to generalize it in future versions.

**Review by NicolasHug — COMMENTED** [review]
Here are my initial comments @fmassa @datumbox @haooooooqi !

File: torchvision/models/optical_flow/_raft/raft.py
Comment (datumbox) [reply torchvision/models/optical_flow/_raft/raft.py:12]: FYI: Agreed. One more reason why we shouldn't expose this on the `__all__`.

File: torchvision/models/optical_flow/_raft/raft.py
Comment (datumbox) [inline torchvision/models/optical_flow/_raft/raft.py:111]: Nit: RegNets and ViT started adding these in a private method (`_reset_parameters` or `_init_weights`).

Should we start a convention and align the two implementations? cc @kazhang @sallysyw

**Review by datumbox — COMMENTED** [review]
@NicolasHug Great work.

I'm reviewing code practices and API, not ML validity as this is something you checked already with @haooooooqi.

I clearly marked my comments as FYI (which are just for discussion), Questions (where I just want more info), NIT (which are non-blocking and could optionally be addressed on follow up PRs) and Blocking (which I think need to be done here).

My Blocking comments are minimal as you can see below and can be addressed easily. After addressing them, ping me to approve but I would wait for the review of @fmassa prior merging.

File: torchvision/models/optical_flow/_raft/raft.py
Comment (fmassa) [reply torchvision/models/optical_flow/_raft/raft.py:12]: Let's keep this here as is

File: torchvision/models/optical_flow/_raft/raft.py
Comment (fmassa) [reply torchvision/models/optical_flow/_raft/raft.py:151]: I'm ok with this.

File: torchvision/models/optical_flow/_raft/raft.py
Comment (fmassa) [reply torchvision/models/optical_flow/_raft/raft.py:111]: We can make everything consistent, but this can probably be tackled as a separate task that addresses this globally in torchvision

File: torchvision/models/optical_flow/_raft/raft.py
Comment (fmassa) [reply torchvision/models/optical_flow/_raft/raft.py:185]: We already have `torch.cat` in the implementation, and having to have 2 different `nn.Sigmoid` + a `nn.Tanh` will be annoying.

I believe the new FX-based quantization doesn't have those limitations, and as we are not planning on quantizing this model as of now, I would say to keep this as is. IMO it reads better, and using the functional API for operations that don't hold state was the "recommended" way at some point.

File: torchvision/models/optical_flow/_raft/raft.py
Comment (fmassa) [reply torchvision/models/optical_flow/_raft/raft.py:377]: I still hold to my previous comment, although I have to say that `torch.relu` being visible to the user is probably a mistake (an artifact from how operators were registered in PyTorch). I think we should use instead `torch.nn.functional.relu`.

File: torchvision/models/optical_flow/_raft/raft.py
Comment (NicolasHug) [reply torchvision/models/optical_flow/_raft/raft.py:475]: Sounds good, I'll change to pretrained to avoid any risk.

As long as we graduate the new model builders from prototype to main, we should be able to change back `pretrained` into `weights`, right?

File: torchvision/models/optical_flow/_raft/raft.py
Comment (NicolasHug) [reply torchvision/models/optical_flow/_raft/raft.py:111]: That makes sense, I created a `_init_weights()` method

File: torchvision/models/optical_flow/_raft/raft.py
Comment (datumbox) [reply torchvision/models/optical_flow/_raft/raft.py:475]: That is correct. There are also a bunch of stuff we will need to create for this model to introduce it to the prototype builders (such as the presets used for inference) but this can be addressed separately on another PR. Can help you with that.

File: torchvision/models/optical_flow/__init__.py
Comment (NicolasHug) [reply torchvision/models/optical_flow/__init__.py:1]: Thanks for the feedback 

> Can you talk about the choice to create a second private submodule _raft? Why is that?

I wrote it like this out of habit but I had no intention of keeping it this way

Instead of renaming `raft()` to `raft_large()`, would it be OK instead to rename the `.py` file to something like `raft_core.py` ? Or `raft_implem.py`?

We would have `raft()`, `raft_small` and `RAFT` available from `torchvision.models.optical_flow` and then we would be able to access the other building blocks like `ResidualBlock` in `torchvision.models.optical_flow.raft_core`  -- I would make sure to exclude them from `__all__`.

Would that be OK?

File: torchvision/models/optical_flow/_raft/utils.py
Comment (fmassa) [inline torchvision/models/optical_flow/_raft/utils.py:6]: Food for thought: would it be a significant limitation if we were to return the flows in normalized coordinates, and perform all computations in normalized coordinates?

Or is it standard that flow images are in absolute coordinates?

Some of the benefits of keeping it in relative coordinates is that you don't need to multiply by the scaling factor when upsampling an image.

**Review by fmassa — APPROVED** [review]
Thanks a lot @NicolasHug !
Approving to unblock, but check @datumbox blocking comments as well before merging.

File: torchvision/models/optical_flow/__init__.py
Comment (datumbox) [reply torchvision/models/optical_flow/__init__.py:1]: Though calling it `raft_large` would make it consistent with other small/large models, I can see that on the paper they don't actually call it large. They do use `Raft-s` or "small" for the small version, so that's canonical at least. So I understand why you want to avoid naming it like that.

Renaming the file should be OK but @pmeier should confirm. Though I'm not sure what the actual name should be. We typically name the files after their algorithm and we let the model builder have extra info on the variant. 

@fmassa thoughts on this?

Comment (NicolasHug) [reply torchvision/models/optical_flow/__init__.py:1]: OK, I'll use `raft.py` and `raft_large()` for now to move forward. We can revisit later if needed, thanks for the input

File: torchvision/models/optical_flow/_raft/raft.py
Comment (NicolasHug) [reply torchvision/models/optical_flow/_raft/raft.py]: I'll see if it helps w.r.t. JIT, but otherwise I feel like we don't really need to make this a `nn.Module`. As you said the semantics don't match right. The original code uses `__call__` instead of `index_pyramid`, but I deliberately changed the name because I thought it would be clearer what the function actually does.

File: torchvision/models/optical_flow/raft.py
Comment (datumbox) [inline torchvision/models/optical_flow/raft.py:483]: Perhaps using `NotImplementedError` is indeed the right exception here but I would recommend for now to throw:
```suggestion
        raise ValueError(f"No checkpoint is available for model.")
```

https://github.com/pytorch/vision/blob/3d8723d5f7782881fc19d802819130857ed68ca4/test/test_prototype_models.py#L45-L46



This is because some tests check for this specific exception:

File: test/test_models.py
Comment (NicolasHug) [inline test/test_models.py:823]: I'm parametrizing over this because testing with `_check_jit_scriptable` unfortunately fails on very few entries, e.g.:

```
Mismatched elements: 153 / 11520 (1.3%)
Greatest absolute difference: 0.0002608299255371094 at index (0, 0, 79, 45) (up to 0.0001 allowed)
Greatest relative difference: 0.021354377198448304 at index (0, 1, 53, 68) (up to 0.0001 allowed)
```

I could add `tol` parameters to the check, but I feel like this current test is just as fine

**Review by datumbox — APPROVED** [review]

**Comment by NicolasHug** [comment]
Test failures are unrelated, merging. Thanks a lot for the reviews!!

**Pull request merged**
